cohortextractor/tpp_backend.py

```diff
@@ -2720,6 +2721,92 @@ def patients_outpatient_appointment_date(
               OPA.Patient_ID
             """
 
+    def patients_with_value_from_file(self, f_path, returning=None, type_=None):
+        if not is_csv_filename(f_path):
+            raise TypeError(f"Unexpected file type {f_path}")
+
+        if returning is not None and type_ is None:
+            # StudyDefinition.get_pandas_csv_args() should have raised an error before
+            # we reach here.
+            raise TypeError("If `returning` is passed, then `type_` must be passed")
+
+        if returning is None and type_ is not None:
+            raise TypeError("If `type_` is passed, then `returning` must be passed")
+
+        if returning is None and type_ is None:
+            # If we set these here, then we don't need to keep track of which is None
+            # later.
+            returning = "value"
+            type_ = "bool"
+
+        # Fail before reading the CSV file
+        if type_ == "bool":
+            column_type = "INT"
+        elif type_ == "date":
+            column_type = "DATE"
+        elif type_ == "str":
+            column_type = "VARCHAR(MAX)"
+        elif type_ == "int":
+            column_type = "INT"
+        elif type_ == "float":
+            column_type = "FLOAT"
+        else:
+            # StudyDefinition.get_pandas_csv_args() should have raised an error before
+            # we reach here.
+            raise TypeError(f"Unexpected type {type_}")
+
+        # Which columns of the CSV file should we read?
+        usecols = ["patient_id"]
+        if returning != "value":
+            usecols.append(returning)
+
+        # Read the CSV file into a data frame
+        values = pandas.read_csv(
+            f_path, index_col="patient_id", usecols=usecols, dtype=str
+        )
+
+        if returning == "value":
+            # Adding a dummy column here makes it easier to generate the SQL
+            values[returning] = 1
+
+        # Generate the SQL
+        table_name = self.get_temp_table_name("file")
+        column_name = self._current_column_name
```

Comment:
9a987cb replaces references to `column_name` with references to `returning`. Should I also be sanitizing the value of `returning` e.g. using `quote`? I think the answer is "Probably not", because the value would come from a user who is allowed to execute code against a backend; not an unknown user, in other words.

Reply:
Ah, this is a good point. I think the general rule should be that no input from the user should be able to result in invalid SQL, even though the users in this case are necessarily trusted so we do need do to something with `returning` here. (And this was true before the most change as well, as we still have to use `returning` as the final column name for everything to work properly.)

But I don't think we can just `quote()` it here, because it's an identifier not a string literal. I think instead we need to check that it contains only alphanumeric characters and underscore, which I think we can do with the `\w` regex class and the ASCII flag:
https://docs.python.org/3.9/library/re.html#index-32

If it doesn't match then we throw an error.

Reply:
Nice! 7a386a8 sanitizes the `returning` kwarg using your suggested approach.